File: partiql-ast/src/main/java/org/partiql/ast/v1/SelectPivot.java
package org.partiql.ast.v1;

import org.jetbrains.annotations.NotNull;
import org.partiql.ast.v1.expr.Expr;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

/**
 * TODO docs, equals, hashcode
 */
public class SelectPivot extends Select {
    @NotNull
    public Expr key;

    @NotNull
    public Expr value;

    public SelectPivot(@NotNull Expr key, @NotNull Expr value) {
        this.key = key;
        this.value = value;
    }

    @NotNull
    @Override
    public Collection<AstNode> children() {
        List<AstNode> kids = new ArrayList<>();
        kids.add(key);
        kids.add(value);
        return kids;
    }

    @Override
    public <R, C> R accept(@NotNull AstVisitor<R, C> visitor, C ctx) {
        return visitor.visitSelectPivot(this, ctx);
    }
}

File: partiql-ast/src/main/kotlin/org/partiql/ast/v1/SelectPivot.kt
package org.partiql.ast.v1

import org.partiql.ast.v1.expr.Expr

/**
 * TODO docs, equals, hashcode
 */
public class SelectPivot(
    @JvmField
    public var key: Expr,
    @JvmField
    public var `value`: Expr,
) : Select() {
    public override fun children(): Collection<AstNode> {
        val kids = mutableListOf<AstNode?>()
        kids.add(key)
        kids.add(value)
        return kids.filterNotNull()
    }

    public override fun <R, C> accept(visitor: AstVisitor<R, C>, ctx: C): R =
        visitor.visitSelectPivot(this, ctx)
}
